### ISSUE TITLE
All detected bulbs have the same name

### ISSUE THREAD

**Issue by timothyb89** [timothyb89/lifx-tasker#2]
When multiple bulbs are on the network, icons for each are shown, but they all seem to have (at least) the same name, and may be duplicates. Unfortunately I only have one bulb at the moment so I'm not able to reproduce.

Not sure if this is an issue upstream in lifx-java or in the UI. Logcat output of the detection process would be invaluable here!

**Comment by Na3blis** [comment]
I can confirm that this does happen. I have 4 bulbs in my room (the only 5 more downstairs but off at the moment) and they all say 'Master Back Right' for the name.

**Comment by Na3blis** [comment]
Not sure how to get the logcat of the detection, here's some from trying to edit though..

```
05-23 23:18:16.671 I/EditActivity(22950): Validation of BULB_NAMES passed
05-23 23:18:16.671 I/EditActivity(22950): action = Power Off
05-23 23:18:16.671 I/EditActivity(22950): bulbs = [Master Back Right, Master Back Right, Master Back Right, Master Back Right]
05-23 23:18:17.481 D/YouTube MDX(13200): discoveryRequestChanged : DiscoveryRequest{ selector=MediaRouteSelector{ controlCategories=[com.google.cast.CATEGORY_CAST] }, activeScan=true, isValid=true }
05-23 23:18:18.131 D/dalvikvm(3654): GC_EXPLICIT freed 2560K, 19% free 25737K/31424K, paused 1ms+5ms, total 49ms
05-23 23:18:18.761 I/o*.t*.l*.t*.LIFXService(22950): LIFX service started
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950): Unable to start listening
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950): java.net.BindException: bind failed: EADDRINUSE (Address already in use)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at libcore.io.IoBridge.bind(IoBridge.java:89)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at java.net.PlainDatagramSocketImpl.bind(PlainDatagramSocketImpl.java:68)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at java.net.DatagramSocket.bind(DatagramSocket.java:457)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at java.nio.DatagramChannelImpl$DatagramSocketAdapter.bind(DatagramChannelImpl.java:536)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.net.BroadcastListener.startListen(BroadcastListener.java:78)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.net.BroadcastListener.startListen(BroadcastListener.java:100)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.tasker.LIFXService.findBulb(LIFXService.java:198)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.tasker.LIFXService.turnOff(LIFXService.java:312)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.tasker.ReceiverService.process(ReceiverService.java:85)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.tasker.ReceiverService_.access$101(ReceiverService_.java:16)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.tasker.ReceiverService_$2.execute(ReceiverService_.java:52)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at org.androidannotations.api.BackgroundExecutor$Task.run(BackgroundExecutor.java:302)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.FutureTask.run(FutureTask.java:237)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at java.lang.Thread.run(Thread.java:841)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950): Caused by: libcore.io.ErrnoException: bind failed: EADDRINUSE (Address already in use)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at libcore.io.Posix.bind(Native Method)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at libcore.io.ForwardingOs.bind(ForwardingOs.java:40)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    at libcore.io.IoBridge.bind(IoBridge.java:87)
05-23 23:18:18.791 E/o*.t*.l*.t*.LIFXService(22950):    ... 18 more


05-23 23:18:52.601 I/ActivityManager(816): START u0 {act=com.twofortyfouram.locale.intent.action.EDIT_SETTING cmp=org.timothyb89.lifx.tasker/.EditActivity_ (has extras)} from pid 3654
05-23 23:18:52.631 I/EditActivity(22950): Got initial parameters...
05-23 23:18:52.631 I/EditActivity(22950): action = Power Off
05-23 23:18:52.631 I/EditActivity(22950): bulbs = [Master Back Right, Master Back Right, Master Back Right, Master Back Right]
05-23 23:18:52.661 I/EditActivity(22950): actionSelected(true, Power Off)
05-23 23:18:52.661 I/EditActivity(22950): Creating editor for BULB_NAMES
05-23 23:18:52.681 D/android.widget.GridLayout(22950): horizontal constraints: x3-x0>=1080, x3-x2<=138, x2-x1<=192, x1-x0<=117 are inconsistent; permanently removing: x3-x2<=138. 
05-23 23:18:52.701 I/ActivityManager(816): Displayed org.timothyb89.lifx.tasker/.EditActivity_: +103ms
05-23 23:18:55.831 I/ActivityManager(816): START u0 {cmp=org.timothyb89.lifx.tasker/.editor.BulbListEditor_ (has extras)} from pid 22950
05-23 23:18:55.841 I/BulbListEditor(22950): Starting service
05-23 23:18:55.901 I/o*.t*.l*.t*.LIFXService(22950): LIFX service started
05-23 23:18:55.901 I/BulbListEditor(22950): LIFX: org.timothyb89.lifx.tasker.LIFXService@42e549d8
05-23 23:18:55.931 D/dalvikvm(22950): GC_CONCURRENT freed 483K, 11% free 19196K/21360K, paused 1ms+5ms, total 27ms
05-23 23:18:55.951 I/ActivityManager(816): Displayed org.timothyb89.lifx.tasker/.editor.BulbListEditor_: +117ms
05-23 23:18:55.951 I/BulbListEditor(22950): Bulbs updated.
05-23 23:18:57.481 D/YouTube MDX(13200): discoveryRequestChanged : DiscoveryRequest{ selector=MediaRouteSelector{ controlCategories=[com.google.cast.CATEGORY_CAST] }, activeScan=true, isValid=true }
05-23 23:19:00.321 I/EditActivity(22950): Validation of BULB_NAMES passed
05-23 23:19:00.321 I/EditActivity(22950): action = Power Off
05-23 23:19:00.321 I/EditActivity(22950): bulbs = [Master Back Right, Master Back Right]
05-23 23:19:00.431 W/InputMethodManagerService(816): Starting input on non-focused client com.android.internal.view.IInputMethodClient$Stub$Proxy@434eb8a8 (uid=10083 pid=22950)
05-23 23:19:01.721 D/dalvikvm(3654): GC_EXPLICIT freed 1836K, 19% free 25758K/31424K, paused 2ms+3ms, total 52ms
05-23 23:19:02.281 I/o*.t*.l*.t*.LIFXService(22950): LIFX service started
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950): Unable to start listening
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950): java.net.BindException: bind failed: EADDRINUSE (Address already in use)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at libcore.io.IoBridge.bind(IoBridge.java:89)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at java.net.PlainDatagramSocketImpl.bind(PlainDatagramSocketImpl.java:68)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at java.net.DatagramSocket.bind(DatagramSocket.java:457)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at java.nio.DatagramChannelImpl$DatagramSocketAdapter.bind(DatagramChannelImpl.java:536)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.net.BroadcastListener.startListen(BroadcastListener.java:78)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.net.BroadcastListener.startListen(BroadcastListener.java:100)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.tasker.LIFXService.findBulb(LIFXService.java:198)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.tasker.LIFXService.turnOff(LIFXService.java:312)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.tasker.ReceiverService.process(ReceiverService.java:85)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.tasker.ReceiverService_.access$101(ReceiverService_.java:16)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at org.timothyb89.lifx.tasker.ReceiverService_$2.execute(ReceiverService_.java:52)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at org.androidannotations.api.BackgroundExecutor$Task.run(BackgroundExecutor.java:302)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.FutureTask.run(FutureTask.java:237)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at java.lang.Thread.run(Thread.java:841)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950): Caused by: libcore.io.ErrnoException: bind failed: EADDRINUSE (Address already in use)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at libcore.io.Posix.bind(Native Method)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at libcore.io.ForwardingOs.bind(ForwardingOs.java:40)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    at libcore.io.IoBridge.bind(IoBridge.java:87)
05-23 23:19:02.301 E/o*.t*.l*.t*.LIFXService(22950):    ... 18 more
```

**Comment by timothyb89** [comment]
Looks like that particular error is being caused by another LIFX app in the background keeping us from scanning for bulbs. I'll see if I can add a warning instead of just erroring out in the background, but unfortunately its a separate issue.

The official LIFX app is the usual culprit since it likes to hold on to the port and not let go until the app is killed. You can try force closing it, or rebooting if its still causing trouble.

If you get a chance, can you post the logcat output again after killing any other LIFX apps? Ideally you should see something like this:

```
I/SimpleControlActivity( 2319): Starting service
I/o*.t*.l*.t*.LIFXService( 2319): LIFX service started
I/SimpleControlActivity( 2319): LIFX: org.timothyb89.lifx.tasker.LIFXService@41dacb80
I/o*.t*.l*.t*.LIFXService( 2319): Found gateway: Gateway(ipAddress=/192.168.0.105:56700, port=56700, macAddress=MACAddress(hex=D0:73:D5:00:E7:73), bulbs=[])
I/SimpleControlActivity( 2319): Bulbs updated.
I/ActivityManager( 2319): Timeline: Activity_idle id: android.os.BinderProxy@41d2b7e0 time:80066
I/o*.t*.l*.t*.LIFXService( 2319): Found bulb: Bulb(address=MACAddress(hex=D0:73:D5:00:E7:73), powerState=OFF, label=Bedside Lamp)
I/SimpleControlActivity( 2319): Bulbs updated.
```

It's the `Found bulb: Bulb(...)` messages in particular that are of interest. You should see them when opening the app initially.

**Comment by Na3blis** [comment]
Didn't seem to start up discovery until I actually went in to edit the task. But it does appear to pick up the labels properly.

```
05-24 00:06:45.292 I/ActivityManager(819): Start proc com.lifx.lifx for broadcast com.lifx.lifx/.connection.WifiReceiver: pid=4613 uid=10153 gids={50153, 3003}
05-24 00:07:07.632 I/ActivityManager(819): Killing 4613:com.lifx.lifx/u0a153 (adj 15): empty #17
05-24 00:07:29.672 I/ActivityManager(819): START u0 {act=com.twofortyfouram.locale.intent.action.EDIT_SETTING cmp=org.timothyb89.lifx.tasker/.EditActivity_ (has extras)} from pid 4174
05-24 00:07:29.692 I/ActivityManager(819): Start proc org.timothyb89.lifx.tasker for activity org.timothyb89.lifx.tasker/.EditActivity_: pid=5703 uid=10083 gids={50083, 3003}
05-24 00:07:30.002 I/ActivityManager(819): Displayed org.timothyb89.lifx.tasker/.EditActivity_: +316ms
05-24 00:07:31.172 I/ActivityManager(819): START u0 {cmp=org.timothyb89.lifx.tasker/.editor.BulbListEditor_ (has extras)} from pid 5703
05-24 00:07:31.272 I/o*.t*.l*.t*.LIFXService(5703): LIFX service started
05-24 00:07:31.302 I/ActivityManager(819): Displayed org.timothyb89.lifx.tasker/.editor.BulbListEditor_: +126ms
05-24 00:07:31.302 I/BulbListEditor(5703): LIFX: org.timothyb89.lifx.tasker.LIFXService@42bb3798
05-24 00:07:31.662 I/o*.t*.l*.t*.LIFXService(5703): Found gateway: Gateway(ipAddress=/192.168.11.50:56700, port=56700, macAddress=MACAddress(hex=D0:73:D5:00:BE:02), bulbs=[])
05-24 00:07:31.882 I/o*.t*.l*.t*.LIFXService(5703): Found bulb: Bulb(address=MACAddress(hex=D0:73:D5:00:37:A0), powerState=OFF, label=Master Front Left)
05-24 00:07:32.002 I/o*.t*.l*.t*.LIFXService(5703): Found bulb: Bulb(address=MACAddress(hex=D0:73:D5:00:B7:59), powerState=OFF, label=Master Front Right)
05-24 00:07:32.112 I/o*.t*.l*.t*.LIFXService(5703): Found bulb: Bulb(address=MACAddress(hex=D0:73:D5:00:AB:8B), powerState=OFF, label=Master Back Right)
05-24 00:07:32.292 I/o*.t*.l*.t*.LIFXService(5703): Found bulb: Bulb(address=MACAddress(hex=D0:73:D5:00:BE:02), powerState=OFF, label=Master Back Left)
05-24 00:07:51.392 I/ActivityManager(819): START u0 {cmp=org.timothyb89.lifx.tasker/.editor.BulbListEditor_ (has extras)} from pid 5703
05-24 00:07:51.462 I/o*.t*.l*.t*.LIFXService(5703): LIFX service started
05-24 00:07:51.462 I/BulbListEditor(5703): LIFX: org.timothyb89.lifx.tasker.LIFXService@42bb3798
05-24 00:07:51.502 I/ActivityManager(819): Displayed org.timothyb89.lifx.tasker/.editor.BulbListEditor_: +105ms
05-24 00:11:00.226 E/AppsCustomizePagedView(1102): Widget ComponentInfo{com.practice.lifxcontrolwidget.app/com.practice.lifxcontrolwidget.app.provider.LIFXAppWidgetProviderHorizontalStyle} can not fit on this device (1200, 300)
```

**Comment by timothyb89** [comment]
Alright, good to know, thanks! That's the exact output I needed. Looks like it's a UI issue which should hopefully be an easy fix, I'll start investigating.

**Comment by timothyb89** [comment]
I made a quick build that might fix the issue. If you get an opportunity, could you give this apk a try? https://drive.google.com/file/d/0B30rtS9KioFNMjMwVGpwaDVGUkk/edit?usp=sharing

**Comment by Na3blis** [comment]
Nope :(. I do see the Error popup about another program using it though. It shows up every time I tell the task to run, and I haven't opened any other up, just using tasker. But still shows same name for all 4 bulbs. Sent you the logfile to the email on your site

**Comment by timothyb89** [comment]
Thanks for the logs!

It looks like you've got another widget running in the background that's also using the port we need (`com.practice.lifxcontrolwidget.app`) and causing that error. I don't know that there's much I'm able to do about it, but I'll look into ways to at least show what app isn't playing nice to help with debugging in the future.

I'll continue investigating the duplicate labels, but if you're able to get actions working with that widget disabled, could you report back with which of the bulbs is triggered when you try to run the action? Particularly if it's all of them, none of them, the correct one, some random bulb ... any oddness you notice could be a big help!

**Comment by Na3blis** [comment]
Ok, uninstalled that app now. But I'm still getting that error and no
actions are working. Once I get an action working I'll report back on
what/how many lights are working

**Comment by timothyb89** [comment]
Examining your logs I noticed a number of issues and attempted to correct them (a couple of possible race conditions) and added some additional logging.

https://drive.google.com/file/d/0B30rtS9KioFNVnJHbXBaRUZwOEE/edit?usp=sharing

I believe an issue causing the port in use / bind errors was a bug and should hopefully be fixed. It seems that in some situations it would try to bind twice if the service was killed, restarted, and used to execute bulb commands quickly  - which would cause the error along with a few other issues. Hopefully this build should be in improvement if you'd like to give it a spin!

I'm still investigating the duplicate bulb names but it seems that it was part of a larger issue which is getting closer to being resolved. I'll continue work on it tomorrow, but if possible, logcat output from this build would be fantastic!

Thanks again for testing things out, it's a massive help!

**Comment by timothyb89** [comment]
Alright, prepared another test build. I'm pretty sure this should correct the duplicate bulbs - there was a bug in lifx-java that would cause bulb info to get overwritten by other bulbs. I'm not sure if this will get actions working but it should be a step in the right direction (this was definitely a contributing factor).

If you'd like, grab it here: https://drive.google.com/file/d/0B30rtS9KioFNR2FSbXJqOFliVkU/edit?usp=sharing

Thanks again for testing!

**Comment by timothyb89** [comment]
Another test build: https://drive.google.com/file/d/0B30rtS9KioFNNVJMLVFlazZjdk0/edit?usp=sharing

I rewrote a lot of the action handling code. Can't say whether or not it's fixed, but it's worth a shot.

**Comment by timothyb89** [comment]
Apologies for the delays, hit a bit of a block in solving the actions issue, but for now I'm going to mark this bug as closed and continue work in bug #1. I've gotten a couple of confirmations that this latest test build fixed the duplicate issue so I believe this issue is solved.
